### ISSUE TITLE
Error reading far-away sensors

### ISSUE THREAD

**Issue by geekdave** [xperimental/flowercare-exporter#1]
Thanks for putting this out there!  I'm trying to use your exporter instead of a much more complicated setup involving an MQTT bridge and NodeRed using https://github.com/ThomDietrich/miflora-mqtt-daemon.  Using that setup was working, but sometimes it was several minutes between updates, because my sensors are outside in a garden and my Raspberry Pis are inside the house.

When I try with your tool I see this for all my sensors:

```
Error updating "FrontPorchFlowers (C4:7C:8D:63:CD:BE)": can not read firmware: exit status 1
```

I notice you have a dependency on https://github.com/barnybug/miflora and in the readme they mention that retries are not handled by the library.  I wonder if retries could be necessary for far-away sensors that may take a few minutes to be seen?

I'm pretty new to this LE stuff so I might be way off base... just wanted to hopefully get your exporter working since it's exactly what I wanted!

Thanks again.

**Comment by xperimental** [comment]
Thanks for the report @geekdave . I am pretty sure that the currently implementation is still lacking if you have many sensors or the connection is bad. I noticed this as well as I recently increased the number of sensors I want to monitor.

I am currently trying to move the Bluetooth communication away from the `gatttool` implementation and, if possible, to something that is implemented in Go directly. I want to do this mostly, so that I can monitor the bluetooth device continuously instead of running the `gatttool` periodically. There are already a few BLE implementations around, but so far I did not have any luck getting those to work with the sensors.

A recent release also changed that the sensors are not polled in parallel but in sequence (if you have more than one). My theory was that this should reduce errors, but instead the errors increased. I suspect the power source or the cooling of the RPi3 that I am running the exporter on to be insufficient.

That said, the exporter will automatically retry getting data from the sensors every "refresh interval", so even if there are sporadic errors you should still see data at some point.

**Comment by xperimental** [comment]
Sorry for the long silence, I had forgotten about this issue and also deactivated my Flowercare sensors for a while.

I've replaced the existing implementation using `gatttool` with an implementation directly using BLE from Go (using [go-ble](https://github.com/go-ble/ble)) and this seems to work nicely with my two sensors so far. Is it possible for you to build the current version yourself and test with that? I still have one rework in mind which should make the retries better which I'd like to do before releasing a new version.

**Comment by xperimental** [comment]
I've just merged the changes regarding the polling queue (#2) and created a new release (v0.3.0).

For me this seems to fix the issues I was having before with sensors not getting read correctly, but I don't have that many sensors, so it would be nice to have some other opinion on this.

I am using this query to check if there are issues with one or more sensors:

```
avg(avg_over_time((time() - flowercare_updated_timestamp)[60m:])) by (name)
```

It should be a "wavy line" for each sensor hovering around 50% of the refresh interval.

**Comment by xperimental** [comment]
I have been running this fine for a while now and consider this fixed. I hope it is fixed for you too.
